plugins/pathable.cpp

```diff
@@ -203,24 +208,105 @@ struct FloodCtx {
+        return true;
+    // NOTE: smoothing a boulder turns it into a smoothed floor
+    else if (shape == df::tiletype_shape::STAIR_UP || shape == df::tiletype_shape::STAIR_DOWN ||
+            shape == df::tiletype_shape::STAIR_UPDOWN || shape == df::tiletype_shape::BOULDER ||
+            shape == df::tiletype_shape::EMPTY || shape == df::tiletype_shape::NONE)
+        return false;
+    else if (special == df::tiletype_special::TRACK)
+        return false;
+    // Fires seem to have their own path group, and group for lava is 0
+    // According to wiki, the wagon won't path thru pool and river tiles, but ramps are ok
+    else if (material == df::tiletype_material::POOL || material == df::tiletype_material::RIVER)
+        return false;
+
+    return true;
+}
+
+static bool is_wagon_traversible(FloodCtx & ctx, const df::coord & pos, const df::coord & prev_pos) {
     auto tt = Maps::getTileType(pos);
     if (!tt)
         return false;
 
     auto shape = tileShape(*tt);
-    if (shape == df::tiletype_shape::STAIR_UP || shape == df::tiletype_shape::STAIR_UPDOWN)
-        return false;
+    auto& occ = *Maps::getTileOccupancy(pos);
+    switch (occ.bits.building) {
+        case tile_building_occ::Obstacle: // Statues, windmills (middle tile)
+            //FALLTHROUGH
+        case tile_building_occ::Well:
+            //FALLTHROUGH
+        case tile_building_occ::Impassable: // Raised bridges
+            return false;
+
+        case tile_building_occ::Dynamic:
+            // doors(block), levers (block), traps (block), hatches (OK, but block on down ramp)
+            // closed floor grates (OK), closed floor bars (OK)
+            if (is_wagon_dynamic_traversible(shape, pos) == false)
+                return false;
+            break;
+
+        case tile_building_occ::None: // Not occupied by a building
+            //FALLTHROUGH
+        case tile_building_occ::Planned:
+            //FALLTHROUGH
+        case tile_building_occ::Passable:
+            // Any tile with no building or a passable building including
+            // beds, supports, rollers, armor/weapon stands, cages (not traps),
+            // open wall grate/vertical bars, retracted bridges, open floodgates,
+            // workshops (tiles with open space are handled by the tile check)
+            if (is_wagon_tile_traversible(*tt) == false)
+                return false;
+            break;
+        case tile_building_occ::Floored:
+            // depot, lowered bridges or retractable bridges, forbidden hatches
+            break;
+        default:
+            //NOTREACHED
+            break;
+    }
 
     if (ctx.wgroup == Maps::getWalkableGroup(pos))
         return true;
 
+    // RAMP_TOP is assigned a walkability group if that commit is accepted
+    // so this test, I think, would be useless.
     if (shape == df::tiletype_shape::RAMP_TOP ) {
         df::coord pos_below = pos + df::coord(0, 0, -1);
         if (Maps::getWalkableGroup(pos_below)) {
```

Comment:
assuming the Maps module change does *not* go in quite yet, do we need to add logic here that checks whether the walkability group of `pos_below` matches `ctx.wgroup`?